### DISCUSSION TITLE
Try to run nginx in WasmEdge, but fail

### DISCUSSION THREAD

**Discussion by hungryzzz** [WasmEdge/WasmEdge#2326] [Q&A]
Hi guys, I try to execute nginx in WasmEdge but I get something wrong, followings are my steps to do it:

1. Build nginx(v1.23.3) using [Emscripten](https://emscripten.org/docs/compiling/Building-Projects.html)
- `emconfigure ./configure --without-http_rewrite_module --without-http_gzip_module --with-cc-opt="-Wno-sign-compare"`
- `emmake make`
When I run the second cmd, I meet some errors but solve them by adding two build options(`-sSTANDALONE_WASM -sERROR_ON_UNDEFINED_SYMBOLS=0`) in file `obj/Makefile`.
![1677656089194](https://user-images.githubusercontent.com/32137313/222073957-fafe3e37-b743-430e-ab04-556434997d69.jpg)
After building it successfully, I can find the final .wasm file in `obj/nginx.wasm`.

2. Execute the nginx in WasmEdge using `sudo wasmedge nginx.wasm -v`, but I get the errors as following:
```
[2023-03-01 02:21:49.361] [error] instantiation failed: unknown import, Code: 0x62
[2023-03-01 02:21:49.361] [error]     When linking module: "env" , function name: "getaddrinfo"
[2023-03-01 02:21:49.361] [error]     At AST node: import description
[2023-03-01 02:21:49.361] [error]     This may be the import of host environment like JavaScript or Golang. Please check that you've registered the necessary host modules from the host programming language.
[2023-03-01 02:21:49.361] [error]     At AST node: import section
[2023-03-01 02:21:49.361] [error]     At AST node: module
```

The OS I use is 20.04.1-Ubuntu.
Does anybody know why this error occur? Is there something wrong during the building or do I need to add some other options when I use WasmEdge?
Thank you a lot!

**Comment by hydai** [comment]
Hi, It looks like the `emcc` generated `getaddrinfo` in the module called `env`, which is not supported in the WASI spec. That's why it failed. Does the `emcc` generate other output files including the `env` lib?

**Reply by hungryzzz** [reply]
Hi, I didn't find the output files including the `env` lib. What kind of format should the `env` lib be?

**Reply by hydai** [reply]
I have no idea about `emcc`, maybe `env.wasm` or any file that can be imported by your wasm application file?

**Reply by hungryzzz** [reply]
Thank you very much! I still can't find the `env` lib so I also open an [issue](https://github.com/emscripten-core/emscripten/issues/18878) in Emscripten's repo.

**_"Note that the .wasm file is not standalone - it’s not easy to manually run it without that .js code, as it depends on getting the proper imports that integrate with JS. "_**

I find the above text from [Emscripten's official document](https://emscripten.org/docs/compiling/WebAssembly.html), and I also find a document about [wasm-standalone](https://github.com/emscripten-core/emscripten/wiki/WebAssembly-Standalone) in Emscripten project. 

Therefore, I think maybe it's hard to compile nginx to a standalone `wasm`, such as a C library function which isn't in `WASI` specification, it can't be implemented by `wasm` and can only rely on the other glue code(JS).
